pkg/api/operators/apis/experiment/v1alpha2/util.go

```diff
@@ -137,3 +141,33 @@ func (exp *Experiment) MarkExperimentStatusFailed(reason, message string) {
 	}
 	exp.setCondition(ExperimentFailed, v1.ConditionTrue, reason, message)
 }
+
+func (exp *Experiment) NeedUpdateFinalizers() (bool, []string) {
+	deleted := exp.GetDeletionTimestamp() != nil
+	pendingFinalizers := exp.GetFinalizers()
+	contained := false
+	for _, elem := range pendingFinalizers {
+		if elem == cleanDataFinalizer {
+			contained = true
+			break
+		}
+	}
+
+	if !deleted && !contained {
+		if exp.Spec.RetainHistoricalData {
+			return false, []string{}
+		}
+		finalizers := append(pendingFinalizers, cleanDataFinalizer)
```

Comment:
Why do we need `cleanDataFinalizer` and exclude it at line 166?

Reply:
by k8s [finalizer feature](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers), we can make sure when "kubectl delete experiment xx", the related data in DB deleted before experiment resource deleted in k8s.

Reply:
I see. Thanks.